libcst/tests/test_fuzz.py

```diff
@@ -0,0 +1,136 @@
+hypothesis.settings.register_profile(
+    name="settings-for-unit-tests",
+    print_blob=True,
+    deadline=timedelta(seconds=10),
+    suppress_health_check=[hypothesis.HealthCheck.too_slow],
+)
+hypothesis.settings.load_profile("settings-for-unit-tests")
+# When the test settings stop finding new bugs, you can run `python test_fuzz.py`
+# to find more.  We turn the number of examples up, and skip the initial "reuse"
+# phase in favor of looking for new bugs... but put everything we find in the
+# database so it will be replayed next time we use the normal settings.
+hypothesis.settings.register_profile(
+    name="settings-for-fuzzing",
+    parent=hypothesis.settings.get_profile("settings-for-unit-tests"),
+    max_examples=100_000,
+    phases=(hypothesis.Phase.generate, hypothesis.Phase.shrink),
+)
+
+
+class FuzzTest(unittest.TestCase):
+    """Fuzz-tests based on Hypothesis and Hypothesmith."""
+
+    @hypothesis.given(source_code=from_grammar(start="file_input"))
+    def test_parsing_compilable_module_strings(self, source_code: str) -> None:
+        """The `from_grammar()` strategy generates strings from Python's grammar.
+
+        This has a bunch of weird edge cases because "strings accepted by Cpython"
+        isn't actually the same set as "strings accepted by the grammar", and for
+        a few other weird reasons you can ask Zac about if you're interested.
+
+        Valid values for ``start`` are ``"single_input"``, ``"file_input"``, or
+        ``"eval_input"``; respectively a single interactive statement, a module or
+        sequence of commands read from a file, and input for the eval() function.
+
+        .. warning::
+            DO NOT EXECUTE CODE GENERATED BY THE `from_grammar()` STRATEGY.
+
+            It could do literally anything that running Python code is able to do,
+            including changing, deleting, or uploading important data.  Arbitrary
+            code can be useful, but "arbitrary code execution" can be very, very bad.
+        """
+        self.reject_invalid_code(source_code, mode="exec")
+        tree = libcst.parse_module(source_code)
+        self.assertEqual(source_code, tree.code)
+
+    @hypothesis.given(source_code=from_grammar(start="eval_input").map(str.strip))
+    def test_parsing_compilable_expression_strings(self, source_code: str) -> None:
+        """Much like statements, but for expressions this time.
+
+        We change the start production of the grammar, the compile mode,
+        and the libCST parse function, but codegen is as for statements.
+        """
+        self.reject_invalid_code(source_code, mode="eval")
+        try:
+            tree = libcst.parse_expression(source_code)
+        except Exception:
+            hypothesis.reject()
+        else:
+            self.assertEqual(source_code, libcst.Module([]).code_for_node(tree))
+
+    @unittest.skip("It's surprisingly tricky to round-trip statements")
```

Comment:
I'm curious, what additional stuff is lurking in statement causing us to not round trip properly? Is this something LibCST should be concerned with?

Reply:
I think we've got a bug where trailing whitespace (on subsequent lines) is silently dropped instead of raising an exception. The fix will probably involve reading the whitespace state object once parsing is done to check that we're at the EOF.

Reply:
That doesn't seem to be too big of a deal. Since this is marked @skip already, we can merge and figure out what the deal is later.

Reply:
I can definitely see a few edge cases with parse_statement and parse_expression playing around with this more. You're exactly right about the trailing whitespace being silently dropped. I can tighten up these checks and add more documentation around parse_statement and parse_expression, then re-enable the statement check.

Reply:
`"pass\r#\n"` -> `"pass\n"` is representative, FYI.

As more general advice, the best way to fix up an input is by construction by the `.map(...)` method of the strategy, then by rejection sampling using `.filter(...)` on the strategy, and finally by rejection with `assume(...)` or `reject()`.  That's basically a question of efficiency, but it also affects how examples are displayed in verbose mode.